eng/tools/select-packages/rush-runner.js

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See the @microsoft/rush package's LICENSE file for license information.
+
+// This script searches for package.json files recursively under a given path
+// excluding packages under a node_modules folder and opens them to determine
+// whether they contain a "sdk-archetype" field with the value of "client".
+//
+// If they do, then the package name is recorded and then after all package.json files
+// have been scanned a single string is output using the Azure Pipelines Commands special
+// syntax which creates a variable for use within a later stage of the pipeine.
+
+const fs = require('fs');
+let path = require('path');
+
+function walk(dir, checks) {
```

Comment:
This is probably overkill for what we need. Since we know the repo structure is `sdk/<serviceDir>/<libraryName>/package.json`, we should be able to do something like the following:

```js
const searchDir = path.resolve(`../../../sdk/${serviceDir}`);

const packageJsons = fs.readdirSync(searchDir)
  .filter(f => !f.startsWith('arm-')) // exclude libraries starting with "arm-"
  .map(f => path.join(searchDir, f, 'package.json')) // turn potential directory names into package.json paths
  .filter(f => fs.existsSync(f)); // only keep paths for files that actually exist
``` 

Once you have that, you just need to crack them open and extract both the package names (for providing flags to Rush) as well as package directories (for calling `rushx`). You could do this with more map() and filter() calls, but I feel like that would be a lot harder to read than the iterative solution:

```js
const packageNames = [];
const packageDirs = [];
for (const filePath of packageJsons) {
  const contents = JSON.parse(fs.readFileSync(filePath, 'utf8'));
  if (contents['sdk-type'] === 'client') {
    packageNames.push(contents.name);
    packageDirs.push(path.dirname(filePath);
  }
}
```

Now you have two arrays with all the package names you need to construct calls to `rush`, and all the directories you need to `cd` into in order to run `rushx`.

Reply:
You can combine these together with the logic to search either one service directory, or all service directories, like this:

```js
const getPackageJsons = (serviceDir) => {
  const searchDir = path.resolve(`../../../sdk/${serviceDir}`);

  const packageJsons = fs.readdirSync(searchDir)
    .filter(f => !f.startsWith('arm-')) // exclude libraries starting with "arm-"
    .map(f => path.join(searchDir, f, 'package.json')) // turn potential directory names into package.json paths
    .filter(f => fs.existsSync(f)); // only keep paths for files that actually exist
  
  return packageJsons;
}

let packageJsons = [];
if (serviceDir) {
  packageJsons = getPackageJsons(serviceDir);
} else {
  const sdkDir = path.resolve(`../../../sdk`);

  const serviceDirs = fs.readdirSync(sdkDir)
    .map(f => path.join(sdkDir, f)) // turn directory names into paths
    .filter(f => fs.lstatSync(f).isDirectory()); // only keep those which are actually directories (not files)

  for (const dir of serviceDirs) {
    packageJsons.concat(getPackageJsons(dir));
  }
}

const packageNames = [];
const packageDirs = [];
for (const filePath of packageJsons) {
  const contents = JSON.parse(fs.readFileSync(filePath, 'utf8'));
  if (contents['sdk-type'] === 'client') {
    packageNames.push(contents.name);
    packageDirs.push(path.dirname(filePath);
  }
}
```